### PR TITLE
Add `XML::Reader#close`

### LINKED ISSUE #12649: Parse errors leaking from `XML::Reader`
When reading an invalid document with `XML::Reader`, parser errors remain stored and are falsely attributed to subsequent parses.

Example:
```cr
require "xml"

reader = XML::Reader.new(%(</foo>)) # Invalid XML
reader.read
reader.expand?

xml = XML.parse(%(<people></people>))
xml.errors # => [#<XML::Error:StartTag: invalid element name>]
```

A reason for that is that the `XML::Reader` instance is never closed to clean up. It does not even provide a method for doing that.

And even then there's the problem of global error state in libxml2. If multiple `XML::Reader` instance are intermingled (or an `XML::Reader` with `XML.read`), errors would be totally messed up.

A simple solution for the problem demonstrated above would be to reset errors in `XML.parse` and its sister methods. These methods are self-contained, so the error state could even be reset to what it was before. This would preserver functionality when called within an outer `XML::Reader` "session". 
There's no solution for intermingled `XML::Reader` instances though.

This bug was discovered through spec randomization in https://github.com/crystal-lang/crystal/pull/12541 (https://github.com/crystal-lang/crystal/actions/runs/3303805339/jobs/5452171721).

### PR DESCRIPTION
**Pull request by straight-shoota**
Errors in `XML::Reader` are stored in the global error array at `XML::Error.errors`. They are never cleaned up unless doing so explicitly with `XML::Error.errors.clear`. But there is no mechanism to indicate directly to the reader instance, that it is finished. Because of this, libXML resources are never freed (neither explicitly nor on finalize).

* Adds the method `XML::Reader#close` which signals the reader is finished. It clears the error array and releases libXML resources. This method is used in all specs to prevent leakage into other specs.
* Also adds a finalize handler that releases libXML resources (in case `#close` was not called).
* Also adds `XML::Reader#errors` which delegates to `XML::Error.errors`. This might be debatable because it might be better to *not* disguises the global error state.
* Also adds some general documentation to `XML::Reader`.

Resolves #12649
Based on https://github.com/crystal-lang/crystal/pull/12652

### CODE REVIEW

File: src/xml/reader.cr [hunk 1]
```diff
@@ -170,4 +184,32 @@ class XML::Reader
   def to_unsafe
     @reader
   end
+
+  # Returns the errors reported while parsing. Returns `nil` if there are no errors.
+  #
+  # Errors in the parser implementation are global and this method returns just `XML::Error.errors`.
+  # When mixing
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/xml/reader.cr]: Fixed.

File: src/xml/reader.cr [hunk 2]
```diff
@@ -1,6 +1,36 @@
 require "./libxml2"
 require "./parser_options"
 
+# `XML::Reader` is a low-level XML parser interface for iterating the contents of
+# an XML document.
+#
+# ```cr
```
Comment (Sija) [inline src/xml/reader.cr:7]:
```suggestion
# ```
```

**Comment by Sija** [comment]
I'd suggest adding `#new(&block)` overload ensuring call to `#close` on block exit.

**Comment by asterite** [comment]
Do we know if errors in Nokogiri are also shared? My bet is not. In that case we should research first how they do it.

**Comment by straight-shoota** [comment]
@asterite Good point. It seems nokogiri sets the error function just for every `xmlTextReaderRead`, instead of once for the entire program.

https://github.com/sparklemotion/nokogiri/blob/2882c88c199d8d46df5e3f31cc98b5e0d08ff3bc/ext/nokogiri/xml_reader.c#L530-L532

That should be a more robust solution. Though it probably still doesn't work when used concurrently (with XMLtextReader reading from a stream).

**Comment by straight-shoota** [comment]
Closing in favour of https://github.com/crystal-lang/crystal/pull/12663. With that change there's no need for `#close`.

`#finalize` is extracted to #12664

**Pull request closed without merging**